orttraining/orttraining/test/graph/gradient_graph_builder_test.cc

```diff
@@ -1015,50 +1167,69 @@ TEST(GradientGraphBuilderTest, TrainingSession_WithPipeline) {
   // 100 -> 199: sub 0
   // 200 -> 299: sub 1
   // 300 -> 399: sub 2
-
-  const std::vector<EventsPerBatch> events = {
+  const std::vector<PipelineBatchInfo> expected_plan = {
       // batch 0
-      {{0, 1, 2, 3},
-       {{-1, 100}, {-1, 200}, {-1, 300}, {202, 203}, {104, 105}}},
+      {{{-1, 100}, {-1, 200}, {-1, 300}, {202, 203}, {104, 105}}, -1},
       // batch 1
-      {{4, 5, 6, 7},
-       {{100, 101}, {200, 201}, {300, 301}, {204, 205}, {106, 107}}},
+      {{{100, 101}, {200, 201}, {300, 301}, {204, 205}, {106, 107}}, -1},
       // batch 2
-      {{8, 9, 10, 11},
-       {{101, 102}, {201, 202}, {301, 302}, {206, 207}, {107, 108}}},
+      {{{101, 102}, {201, 202}, {301, 302}, {206, 207}, {107, 108}}, -1},
       // batch 3
-      {{12, 13, 14, 15},
-       {{102, 103}, {203, 204}, {302, 303}, {208, 209}, {108, 109}}},
+      {{{102, 103}, {203, 204}, {302, 303}, {208, 209}, {108, 109}}, -1},
       //batch 4
-      {{16, 17, 18, 19},
-       {{103, 104}, {205, 206}, {303, 304}, {209, 210}, {109, 110}}},
+      {{{103, 104}, {205, 206}, {303, 304}, {209, 210}, {109, 110}}, -1},
       // batch 5
-      {{20, 21, 22, 23},
-       {{105, 106}, {207, 208}, {304, 305}, {210, 211}, {110, 111}}},
+      {{{105, 106}, {207, 208}, {304, 305}, {210, 211}, {110, 111}}, 0},
```

Comment:
>0 [](start = 69, length = 1)

I wonder if it's necessary to maintain the retired_batch_id here? 

for example, batch 6 would reuse batch_id 1, 7 would reuse 2, and the list goes on... 

But this doesn't sees to carry too much information.

I might be naive here, but does a dynamic context pool work here? When batch 0 complete, the thread is returned to the pool, and when batch 5 comes, it automatically gets the freed resource from the pool ?

Reply:
Will update it to general info about retired_batches.

---
In reply to: [395473947](https://github.com/microsoft/onnxruntime/pull/3206#discussion_r395473947) [](ancestors = 395473947)